<launch>

  <arg name="rosbag_name" default="smb_2022-06-03-14-41-30_smb"/>
  <arg name="rosbag_path" default="/home/mads/Downloads/$(arg rosbag_name).bag"/>

  <arg name="camera_calib_path" default="$(find object_detection)/cfg/smb264_camera_model.yaml"/>

  <arg name="optical_frame_name" default='blackfly_right_optical_link'/>
  <arg name="input_camera_name" default="/versavis/cam0" />
  <arg name="undistorted_image_name" default="$(arg input_camera_name)/undistort" />
  <arg name="lidar_topic" default='/rslidar_points'/>
  <arg name="use_image_undistort" default="true"/>


  <arg name="object_detection_classes" default="[0,1,11,25,13,74]"/>

  <param name ="use_sim_time" value="true"/>
    <!--Debayer the image-->
    <!-- <node pkg="image_proc" type="image_proc" name="debayer" ns="$(arg input_camera_name)/" /> -->
  <group if="$(arg use_image_undistort)">
    <node pkg="nodelet" type="nodelet" name="debayer"
          args="standalone image_proc/debayer">
      <remap from="image_raw" to="$(arg input_camera_name)/image_raw" />
      <remap from="image_color" to="$(arg input_camera_name)/image_color" />
      <remap from="image_mono" to="$(arg input_camera_name)/image_mono" />
    </node>

    <!--Undistored the debayered image-->
    <!--https://github.com/ethz-asl/image_undistort-->
    <node name="image_undistort_node" pkg="image_undistort" type="image_undistort_node" output="screen" clear_params="true">
      <param name="input_camera_namespace" value="versavis"/>      
      <param name="input_camera_info_from_ros_params" value = "true"/>    
      <param name="scale" value="1.0"/>
      <param name="output_camera_info_source" value="auto_generated"/>   
      <param name="publish_tf" value="false"/>
      <param name="output_frame" value="$(arg optical_frame_name)"/>
      <rosparam file="$(arg camera_calib_path)"/>
      <remap from="input/image" to="$(arg input_camera_name)/image_color"/>
      <!-- <remap from="output/camera_info" to="$(arg input_camera_name)/camera_info_undistort"/> -->
      <!-- <remap from="output/image" to="$(arg input_camera_name)/undistorted"/> -->
      <remap from="output/image" to="$(arg undistorted_image_name)/image_color_rect"/>
    </node>
  </group>

  <group unless="$(arg use_image_undistort)">
    <node pkg="image_proc" type="image_proc" name="debayer_and_rect" ns="$(arg input_camera_name)/" />
  </group>



  <!--Object detection node-->
  <node name="objectify" pkg="object_detection" type="node.py" output="screen">

    <!--Input related-->
    <!-- <param name="camera_topic" value='$(arg undistorted_image_name)/image'/> -->
    <param name="camera_topic" value='$(arg undistorted_image_name)/image_color_rect'/>
    <param name="camera_info_topic" value='$(arg undistorted_image_name)/camera_info'/>
    <!-- <param name="camera_info_topic" value='$(arg input_camera_name)/camera_info'/> -->
    <param name="lidar_topic" value='$(arg lidar_topic)'/>

    <!--Output related-->
    <param name="object_detection_topic" value='/objects'/>
    <param name="verbose" value='True'/>

    <!--Camera Lidar synchronization related-->
    <param name="camera_lidar_sync_queue_size" value="10"/>
    <param name="camera_lidar_sync_slop" value="0.1"/>

    <!--Config files-->
    <param name="config_dir" value='$(find object_detection)/cfg/'/>

    <!--Point Projector related-->
    <param name="project_config" value='projector_config_SMB264.yaml'/>

    <!--Object detection related-->
    <param name="model" value='yolov5l6'/>
    <param name="device" value='cpu'/>
    <param name="confident" value='0.4'/>
    <param name="iou" value='0.1'/>
    <rosparam param="classes" subst_value="True">$(arg object_detection_classes)</rosparam> >
    <param name="multiple_instance" value='False'/>

    <!--Object localization related-->
    <param name="model_method" value='hdbscan'/>
    <param name="ground_percentage" value='25'/>
    <param name="bb_contract_percentage" value='10.0'/>
  </node>

  <include file="$(find object_visualization)/launch/object_visualization.launch">
    <arg name="object_topic" value="/objects"/>
    <arg name="visualize_all" value='False'/>
    <arg name="only_BB" value='False'/>
  </include>


  <include file="$(find smb_description)/launch/debug.launch">
    <!--<arg name="visualize" value="1"/>-->
  </include>

  <node pkg="rosbag" type="play" name="player" output="screen" args="-d 3 --clock $(arg rosbag_path)" required = "false"/>

</launch>